modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceMinimumRequiredTimeProviderImpl.java

```diff
@@ -0,0 +1,294 @@
+
+    /**
+     * Constructor.
+     */
+    public RebalanceMinimumRequiredTimeProviderImpl(MetaStorageManager metaStorageManager, CatalogService catalogService) {
+        this.metaStorageManager = metaStorageManager;
+        this.catalogService = catalogService;
+    }
+
+    @Override
+    public long minimumRequiredTime() {
+        // Use the same revision to read all the data, in order to guarantee consistency of data.
+        long appliedRevision = metaStorageManager.appliedRevision();
+
+        // Ignore the real safe time, having a time associated with revision is enough. Also, acquiring real safe time would be
+        // unnecessarily more complicated due to handling of possible data races.
+        long metaStorageSafeTime = metaStorageManager.timestampByRevisionLocally(appliedRevision).longValue();
+
+        long minTimestamp = metaStorageSafeTime;
+
+        Map<Integer, Map<Integer, Assignments>> stableAssignments = readAssignments(STABLE_ASSIGNMENTS_PREFIX_BYTES, appliedRevision);
+        Map<Integer, Map<Integer, Assignments>> pendingAssignments = readAssignments(PENDING_ASSIGNMENTS_PREFIX_BYTES, appliedRevision);
+
+        Map<Integer, Long> pendingChangeTriggerRevisions = readPendingChangeTriggerRevisions(
+                PENDING_CHANGE_TRIGGER_PREFIX_BYTES,
+                appliedRevision
+        );
+
+        int earliestCatalogVersion = catalogService.earliestCatalogVersion();
+        int latestCatalogVersion = catalogService.latestCatalogVersion();
+
+        Map<Integer, Integer> tableIdToZoneIdMap = tableIdToZoneIdMap(earliestCatalogVersion, latestCatalogVersion);
+        Map<Integer, NavigableMap<Long, CatalogZoneDescriptor>> allZonesByTimestamp = allZonesByTimestamp(
+                earliestCatalogVersion,
+                latestCatalogVersion
+        );
+        Map<Integer, NavigableMap<Long, CatalogZoneDescriptor>> allZonesByRevision = allZonesByRevision(allZonesByTimestamp);
+        Map<Integer, Long> zoneDeletionTimestamps = zoneDeletionTimestamps(earliestCatalogVersion, latestCatalogVersion);
+
+        for (Map.Entry<Integer, Integer> entry : tableIdToZoneIdMap.entrySet()) {
+            Integer tableId = entry.getKey();
+            Integer zoneId = entry.getValue();
+
+            NavigableMap<Long, CatalogZoneDescriptor> zoneDescriptors = allZonesByTimestamp.get(zoneId);
+            int zonePartitions = zoneDescriptors.lastEntry().getValue().partitions();
+
+            Long pendingChangeTriggerKey = pendingChangeTriggerRevisions.get(tableId);
+
+            // +-1 here ir required for 2 reasons:
+            // - we need timestamp right before deletion, if zone is deleted, thus we must subtract 1;
+            // - we need a "metaStorageSafeTime" if zone is not deleted, without any subtractions.
+            long latestTimestamp = zoneDeletionTimestamps.getOrDefault(zoneId, metaStorageSafeTime + 1) - 1;
+
+            long zoneRevision = pendingChangeTriggerKey == null
```

Comment:
Let's say that zone was created on revision 10, and updated on revision 12. pendingChangeTriggerKey is null.
What will be the value of 'zoneRevision'?

Reply:
I should add such test and assert what we want to see, if it doesn't exist yer.
My assumption is - we should be able see the earliest catalog descriptor for a zone.